### PR TITLE
3. Add workflow manager tests w/ mock

### PR DESCRIPTION
**Pull request by nikhitashanker77**
- Add a test for workflow manager using gomock
- Revamp file structure so tests and mocks are in same folder as tested function
- Remove unnecessary go.mod files.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,3 +31,7 @@ trademarks or logos is subject to and must follow
 [Microsoft's Trademark & Brand Guidelines](https://www.microsoft.com/en-us/legal/intellectualproperty/trademarks/usage/general).
 Use of Microsoft trademarks or logos in modified versions of this project must not cause confusion or imply Microsoft sponsorship.
 Any use of third-party trademarks or logos are subject to those third-party's policies.
+
+## Run and Test
+To run the main use: go run .
```
Comment (MichaelBehm) [inline README.md]: When giving relative path descriptions, let the users know what directory they should be `cd`d into

**Comment by MichaelBehm** [comment]
Not sure if this was on purpose, but this PR isn't merging into Main, it's merging into one of your dev branches

**Comment by nikhitashanker77** [comment]
(A) Mocking functions in GO

Option 1: Use a function pointers that you replace with the test / actual method implementation

Option 2 [Chosen]: Create an interface and use gomock

Downside: create an interface just to be able to test

(B) Test folder structure


1. Say we want to test workflow_factory which is in main (renamed to root in PR)

In the test we need to import both nodes and the root/main folder

If we have a tests folder, we get a "cycle" from the tests to the main package to the nodes package.
Tests->main->nodes
 |--------------------^
package github.com/microsoft/BladeMonRT/main/tests (test)
        imports github.com/microsoft/BladeMonRT
        imports github.com/microsoft/BladeMonRT/nodes: import cycle not allowed
package github.com/microsoft/BladeMonRT/main/tests (test)
        imports github.com/microsoft/BladeMonRT/nodes: import cycle not allowed

If we don't create a separate tests folder and put the test in main/root, the connection between tests and main is not there.

2. Mock has to be at the same path + package as interface otherwise we get an error.
	Mock implements path+packageA.methodName instead of path+packageB.methodName

Some notes 
==========
kubernetes doesn't have a separate test folder
kubernetes/pkg/controller/disruption at master · kubernetes/kubernetes (github.com)

If this file structure is fine, 
	- I will fix the capitalization of methods in this PR because some methods are only capitalized so that they are visible to the test package
       - I will delete the test folder that use the dummy nodes for testing and not mocks

**Comment by nikhitashanker** [comment]
> Not sure if this was on purpose, but this PR isn't merging into Main, it's merging into one of your dev branches

Yes I did this because this PR depends on #2. Will change base after #2 is merged.

File: README.md (hunk 1 above)
Comment (nikhitashanker) [reply README.md]: I see, done.

Comment (nikhitashanker77) [reply README.md]: I see, done.

**Comment by nikhitashanker77** [comment]
> Not sure if this was on purpose, but this PR isn't merging into Main, it's merging into one of your dev branches

Yes I did this because this PR depends on #2. Will change base after #2 is merged.

**Comment by nikhitashanker77** [comment]
1 Can we have single go.mod in root folder?
Yes. Fixed.

2 Does a return statement with the mocking library run the function and then return the value or just return the value?
Just return the value, if we want to run the original implementation of the function + return a value we can specify that in the test.

3 NodeX has parseXML how do we mock that function since it's not in InterfaceNode?
1 [!] Failed option:
Type Implementing multiple interfaces in Go (Golang) - Welcome To Golang By Example
We are able to implement multiple interfaces in GO
NodeX implements InterfaceNode, InterfaceNodeX

InterfaceNodeX contains all the methods in InterfaceNode

When we test with InterfaceNodeX, we use MockInterfaceNodeX

2 This is called patching / monkey patching and is done using the the patch library in python rather than with mocks.
Monkey patching is more difficult in GO -- available library suggested by https://stackoverflow.com/questions/37079225/golang-monkey-patching
 https://github.com/bouk/monkey
-This library is inserting a jump instruction in the binary
-GO=compiled, Python=interpreted

**Review by MichaelBehm — APPROVED** [review]
LGTM

**Pull request merged**